# Some Models
> To avoid copy-pasta

In [ ]:
# default_exp resnet_models

In [ ]:
# hide
import blackhc.project.script
from nbdev.showdoc import *

Appended /home/blackhc/PycharmProjects/bald-ical/src to paths
Switched to directory /home/blackhc/PycharmProjects/bald-ical
%load_ext autoreload
%autoreload 2


In [ ]:
# exports
from dataclasses import dataclass
from typing import Union

import torch
import torch.nn
import torch.optim
from torch import nn as nn
from torch.nn import functional as F
from torchvision import models

from batchbald_redux.consistent_mc_dropout import (
    BayesianModule,
    ConsistentMCDropout,
    ConsistentMCDropout2d,
)
from batchbald_redux.model_optimizer_factory import ModelOptimizer, ModelOptimizerFactory

In [ ]:
# exports

import torch.nn as nn
from torch import Tensor
from torch.hub import load_state_dict_from_url
from torch.nn import functional as F

model_urls = {
    "resnet18": "https://download.pytorch.org/models/resnet18-5c106cde.pth",
    "resnet34": "https://download.pytorch.org/models/resnet34-333f7ec4.pth",
    "resnet50": "https://download.pytorch.org/models/resnet50-19c8e357.pth",
    "resnet101": "https://download.pytorch.org/models/resnet101-5d3b4d8f.pth",
    "resnet152": "https://download.pytorch.org/models/resnet152-b121ed2d.pth",
    "resnext50_32x4d": "https://download.pytorch.org/models/resnext50_32x4d-7cdf4587.pth",
    "resnext101_32x8d": "https://download.pytorch.org/models/resnext101_32x8d-8ba56ff5.pth",
}


def conv3x3(in_planes, out_planes, stride=1, groups=1, dilation=1):
    """3x3 convolution with padding"""
    return nn.Conv2d(
        in_planes,
        out_planes,
        kernel_size=3,
        stride=stride,
        padding=dilation,
        groups=groups,
        bias=False,
        dilation=dilation,
    )


def conv1x1(in_planes, out_planes, stride=1):
    """1x1 convolution"""
    return nn.Conv2d(in_planes, out_planes, kernel_size=1, stride=stride, bias=False)


class BasicBlock(nn.Module):
    expansion = 1

    def __init__(
        self, inplanes, planes, stride=1, downsample=None, groups=1, base_width=64, dilation=1, norm_layer=None
    ):
        super(BasicBlock, self).__init__()
        if norm_layer is None:
            norm_layer = nn.BatchNorm2d
        if groups != 1 or base_width != 64:
            raise ValueError("BasicBlock only supports groups=1 and base_width=64")
        if dilation > 1:
            raise NotImplementedError("Dilation > 1 not supported in BasicBlock")
        # Both self.conv1 and self.downsample layers downsample the input when stride != 1
        self.conv1 = conv3x3(inplanes, planes, stride)
        self.bn1 = norm_layer(planes)
        self.relu = nn.ReLU(inplace=True)
        self.conv2 = conv3x3(planes, planes)
        self.bn2 = norm_layer(planes)
        self.downsample = downsample
        self.stride = stride

    def forward(self, x):
        identity = x

        out = self.conv1(x)
        out = self.bn1(out)
        out = self.relu(out)

        out = self.conv2(out)
        out = self.bn2(out)

        if self.downsample is not None:
            identity = self.downsample(x)

        out += identity
        out = self.relu(out)

        return out


class Bottleneck(nn.Module):
    expansion = 4

    def __init__(
        self, inplanes, planes, stride=1, downsample=None, groups=1, base_width=64, dilation=1, norm_layer=None
    ):
        super(Bottleneck, self).__init__()
        if norm_layer is None:
            norm_layer = nn.BatchNorm2d
        width = int(planes * (base_width / 64.0)) * groups
        # Both self.conv2 and self.downsample layers downsample the input when stride != 1
        self.conv1 = conv1x1(inplanes, width)
        self.bn1 = norm_layer(width)
        self.conv2 = conv3x3(width, width, stride, groups, dilation)
        self.bn2 = norm_layer(width)
        self.conv3 = conv1x1(width, planes * self.expansion)
        self.bn3 = norm_layer(planes * self.expansion)
        self.relu = nn.ReLU(inplace=True)
        self.downsample = downsample
        self.stride = stride

    def forward(self, x):
        identity = x

        out = self.conv1(x)
        out = self.bn1(out)
        out = self.relu(out)

        out = self.conv2(out)
        out = self.bn2(out)
        out = self.relu(out)

        out = self.conv3(out)
        out = self.bn3(out)

        if self.downsample is not None:
            identity = self.downsample(x)

        out += identity
        out = self.relu(out)

        return out


class BayesianResNet(BayesianModule):
    def __init__(
        self,
        block,
        layers,
        num_classes=1000,
        zero_init_residual=False,
        groups=1,
        width_per_group=64,
        replace_stride_with_dilation=None,
        norm_layer=None,
    ):
        super().__init__()

        if norm_layer is None:
            norm_layer = nn.BatchNorm2d
        self._norm_layer = norm_layer

        self.inplanes = 64
        self.dilation = 1
        if replace_stride_with_dilation is None:
            # each element in the tuple indicates if we should replace
            # the 2x2 stride with a dilated convolution instead
            replace_stride_with_dilation = [False, False, False]
        if len(replace_stride_with_dilation) != 3:
            raise ValueError(
                "replace_stride_with_dilation should be None "
                "or a 3-element tuple, got {}".format(replace_stride_with_dilation)
            )
        self.groups = groups
        self.base_width = width_per_group

        # This gets reset when cifar_mod=True below
        self.conv1 = nn.Conv2d(3, self.inplanes, kernel_size=7, stride=2, padding=3, bias=False)
        self.bn1 = norm_layer(self.inplanes)
        self.relu = nn.ReLU(inplace=True)

        # This gets reset when cifar_mod=True
        self.maxpool = nn.MaxPool2d(kernel_size=3, stride=2, padding=1)

        self.layer1 = self._make_layer(block, 64, layers[0])
        self.layer2 = self._make_layer(block, 128, layers[1], stride=2, dilate=replace_stride_with_dilation[0])
        self.layer3 = self._make_layer(block, 256, layers[2], stride=2, dilate=replace_stride_with_dilation[1])
        self.layer4 = self._make_layer(block, 512, layers[3], stride=2, dilate=replace_stride_with_dilation[2])
        self.avgpool = nn.AdaptiveAvgPool2d((1, 1))

        # self.fc = nn.Linear(512 * block.expansion, num_classes)
        self.fc = nn.Sequential(
            ConsistentMCDropout(),
            nn.Linear(512 * block.expansion, 256 * block.expansion),
            nn.ReLU(),
            ConsistentMCDropout(),
            nn.Linear(256 * block.expansion, 256 * block.expansion),
            nn.Linear(256, num_classes),
        )

        for m in self.modules():
            if isinstance(m, nn.Conv2d):
                nn.init.kaiming_normal_(m.weight, mode="fan_out", nonlinearity="relu")
            elif isinstance(m, (nn.BatchNorm2d, nn.GroupNorm)):
                nn.init.constant_(m.weight, 1)
                nn.init.constant_(m.bias, 0)

        # Zero-initialize the last BN in each residual branch,
        # so that the residual branch starts with zeros, and each residual block behaves like an identity.
        # This improves the model by 0.2~0.3% according to https://arxiv.org/abs/1706.02677
        if zero_init_residual:
            for m in self.modules():
                if isinstance(m, Bottleneck):
                    nn.init.constant_(m.bn3.weight, 0)
                elif isinstance(m, BasicBlock):
                    nn.init.constant_(m.bn2.weight, 0)

    def _make_layer(self, block, planes, blocks, stride=1, dilate=False):
        norm_layer = self._norm_layer
        downsample = None
        previous_dilation = self.dilation
        if dilate:
            self.dilation *= stride
            stride = 1
        if stride != 1 or self.inplanes != planes * block.expansion:
            downsample = nn.Sequential(
                conv1x1(self.inplanes, planes * block.expansion, stride), norm_layer(planes * block.expansion)
            )

        layers = []
        layers.append(
            block(
                self.inplanes, planes, stride, downsample, self.groups, self.base_width, previous_dilation, norm_layer
            )
        )
        self.inplanes = planes * block.expansion
        for _ in range(1, blocks):
            layers.append(
                block(
                    self.inplanes,
                    planes,
                    groups=self.groups,
                    base_width=self.base_width,
                    dilation=self.dilation,
                    norm_layer=norm_layer,
                )
            )

        return nn.Sequential(*layers)

    def deterministic_forward_impl(self, x: Tensor):
        x = self.conv1(x)
        x = self.bn1(x)
        x = self.relu(x)
        x = self.maxpool(x)

        x = self.layer1(x)
        x = self.layer2(x)
        x = self.layer3(x)
        x = self.layer4(x)

        x = self.avgpool(x)
        x = x.reshape(x.size(0), -1)

        return x

    def mc_forward_impl(self, x: Tensor):
        x = self.fc(x)
        x = F.log_softmax(x, dim=-1)
        return x


def _bayesian_resnet(arch, block, layers, cifar_mod, pretrained, progress, **kwargs):
    model = BayesianResNet(block, layers, **kwargs)
    if cifar_mod:
        model.conv1 = torch.nn.Conv2d(3, 64, kernel_size=3, stride=1, padding=1, bias=False)
        model.maxpool = torch.nn.Identity()

    if pretrained:
        assert not cifar_mod
        state_dict = load_state_dict_from_url(model_urls[arch], progress=progress)
        model.load_state_dict(state_dict, strict=False)

    return model


def bayesian_resnet18(*, cifar_mod=False, pretrained=False, progress=True, **kwargs):
    """Constructs a ResNet-18 model.

    Args:
        pretrained (bool): If True, returns a model pre-trained on ImageNet
        progress (bool): If True, displays a progress bar of the download to stderr
    """
    return _bayesian_resnet("resnet18", BasicBlock, [2, 2, 2, 2], cifar_mod, pretrained, progress, **kwargs)


def bayesian_resnet34(*, cifar_mod=False, pretrained=False, progress=True, **kwargs):
    """Constructs a ResNet-34 model.

    Args:
        pretrained (bool): If True, returns a model pre-trained on ImageNet
        progress (bool): If True, displays a progress bar of the download to stderr
    """
    return _bayesian_resnet("resnet34", BasicBlock, [3, 4, 6, 3], cifar_mod, pretrained, progress, **kwargs)

In [ ]:
# exports


@dataclass
class Cifar10BayesianResnetFactory(ModelOptimizerFactory):
    def create_model_optimizer(self) -> ModelOptimizer:
        model = bayesian_resnet18(cifar_mod=True, num_classes=10)
        optimizer = torch.optim.AdamW(model.parameters(), weight_decay=5e-4)
        return ModelOptimizer(model=model, optimizer=optimizer)

## CIFAR-10 40% baseline

In [ ]:
# experiment

from batchbald_redux.datasets import get_dataset
from batchbald_redux.black_box_model_training import train_with_schedule, evaluate

In [ ]:
# experiment 
split_dataset = get_dataset("CIFAR-10", normalize_like_cifar10=False, train_augmentation=True, validation_set_size=5000)

Files already downloaded and verified
Files already downloaded and verified
Files already downloaded and verified


In [ ]:
# experiment

from torch.utils import data

kwargs = {"num_workers": 4, "pin_memory": True}

train_loader = data.DataLoader(split_dataset.train * 0.4, batch_size=128, shuffle=True, **kwargs)
validation_loader = data.DataLoader(split_dataset.validation, batch_size=128, shuffle=False, **kwargs)
test_loader = data.DataLoader(split_dataset.test, batch_size=128, shuffle=False, **kwargs)

model_optimizer = Cifar10BayesianResnetFactory().create_model_optimizer()
train_with_schedule(model=model_optimizer.model, optimizer=model_optimizer.optimizer, train_loader=train_loader, validation_loader=validation_loader, patience_schedule=[3,3,3], factor_schedule=[0.1], max_epochs=60, device="cuda", training_samples=1, validation_samples=10, training_log={})

  2%|1         | 1/60 [00:00<?, ?it/s]

[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.4036, 'crossentropy': 1.7882676845550538}


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.4834, 'crossentropy': 1.481701950454712}


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.5428, 'crossentropy': 1.3699166606903077}


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.501, 'crossentropy': 1.5626290748596192}
Epoch 4: 0.501 worse than 0.5428, patience: 1/3!


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.601, 'crossentropy': 1.2635912508010865}


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.6266, 'crossentropy': 1.187613642501831}


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.6734, 'crossentropy': 1.0052011548995972}


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.682, 'crossentropy': 1.0248640705108643}


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.7096, 'crossentropy': 0.9396376174926758}


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.7272, 'crossentropy': 0.8418712997436524}


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.7102, 'crossentropy': 0.992177451467514}
Epoch 11: 0.7102 worse than 0.7272, patience: 1/3!


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.7522, 'crossentropy': 0.8086827343940735}


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.7696, 'crossentropy': 0.7537960346221924}


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.677, 'crossentropy': 1.2522793543338775}
Epoch 14: 0.677 worse than 0.7696, patience: 1/3!


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.7704, 'crossentropy': 0.7572291868209838}


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.788, 'crossentropy': 0.6703125279903411}


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.8058, 'crossentropy': 0.6643583700180054}


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.7888, 'crossentropy': 0.7042493198394776}
Epoch 18: 0.7888 worse than 0.8058, patience: 1/3!


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.7828, 'crossentropy': 0.7735469184875489}
Epoch 19: 0.7828 worse than 0.8058, patience: 2/3!


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.7842, 'crossentropy': 0.8083664597511292}
Epoch 20: 0.7842 worse than 0.8058, patience: 3/3!


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.7916, 'crossentropy': 0.745104105758667}
Epoch 21: 0.7916 worse than 0.8058, patience: 4/3!
Epoch    21: reducing learning rate of group 0 to 1.0000e-04.


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.8576, 'crossentropy': 0.488191157746315}


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.8572, 'crossentropy': 0.48753206486701967}
Epoch 23: 0.8572 worse than 0.8576, patience: 1/3!


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.8598, 'crossentropy': 0.48893280182480814}


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.863, 'crossentropy': 0.4871544968008995}


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.8594, 'crossentropy': 0.500533379971981}
Epoch 26: 0.8594 worse than 0.863, patience: 1/3!


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.8614, 'crossentropy': 0.4965192899346352}
Epoch 27: 0.8614 worse than 0.863, patience: 2/3!


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.8592, 'crossentropy': 0.5033470409154892}
Epoch 28: 0.8592 worse than 0.863, patience: 3/3!


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.8594, 'crossentropy': 0.504065201023221}
Epoch 29: 0.8594 worse than 0.863, patience: 4/3!
Epoch    29: reducing learning rate of group 0 to 1.0000e-05.


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.8616, 'crossentropy': 0.49568186736106873}
Epoch 30: 0.8616 worse than 0.863, patience: 1/3!


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.8628, 'crossentropy': 0.4986858606278896}
Epoch 31: 0.8628 worse than 0.863, patience: 2/3!


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.8658, 'crossentropy': 0.5025872270286084}


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.8616, 'crossentropy': 0.4966917310476303}
Epoch 33: 0.8616 worse than 0.8658, patience: 1/3!


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.864, 'crossentropy': 0.5035948188602924}
Epoch 34: 0.864 worse than 0.8658, patience: 2/3!


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.864, 'crossentropy': 0.49511414707899093}
Epoch 35: 0.864 worse than 0.8658, patience: 3/3!


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.8628, 'crossentropy': 0.5081136326342821}
Epoch 36: 0.8628 worse than 0.8658, patience: 4/3!


AdamW (
Parameter Group 0
    amsgrad: False
    betas: (0.9, 0.999)
    eps: 1e-08
    lr: 1e-05
    weight_decay: 0.0005
)

In [ ]:
# experiment

evaluate(model=model_optimizer.model, num_samples=20, loader=test_loader, device="cuda")

[1/79]   1%|1          [00:00<?]

{'accuracy': 0.8638, 'crossentropy': 0.5297188091278077}

## CIFAR-100

In [ ]:
# experiment 
split_dataset = get_dataset("CIFAR-100", normalize_like_cifar10=False, train_augmentation=True, validation_set_size=5000)

Files already downloaded and verified


In [ ]:
# experiment

from torch.utils import data

kwargs = {"num_workers": 4, "pin_memory": True}

train_loader = data.DataLoader(split_dataset.train * 0.4, batch_size=128, shuffle=True, **kwargs)
validation_loader = data.DataLoader(split_dataset.validation, batch_size=128, shuffle=False, **kwargs)
test_loader = data.DataLoader(split_dataset.test, batch_size=128, shuffle=False, **kwargs)

model_optimizer = Cifar10BayesianResnetFactory().create_model_optimizer()
train_with_schedule(model=model_optimizer.model, optimizer=model_optimizer.optimizer, train_loader=train_loader, validation_loader=validation_loader, patience_schedule=[3,3,3,3], factor_schedule=[0.1], max_epochs=80, device="cuda", training_samples=1, validation_samples=10, training_log={})

  2%|1         | 1/60 [00:00<?, ?it/s]

[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.037, 'crossentropy': 4.40652554473877}


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.0492, 'crossentropy': 4.2037655502319335}


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.061, 'crossentropy': 4.198613374328613}


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.0564, 'crossentropy': 4.067956605148315}
Epoch 4: 0.0564 worse than 0.061, patience: 1/3!


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.0838, 'crossentropy': 4.026535692596435}


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.0804, 'crossentropy': 4.04006093826294}
Epoch 6: 0.0804 worse than 0.0838, patience: 1/3!


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.0916, 'crossentropy': 3.950498880004883}


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.1032, 'crossentropy': 3.9708983085632323}


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.1142, 'crossentropy': 3.8166026466369627}


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.117, 'crossentropy': 3.8180666912078856}


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.1002, 'crossentropy': 3.947795548248291}
Epoch 11: 0.1002 worse than 0.117, patience: 1/3!


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.1012, 'crossentropy': 3.8586297302246093}
Epoch 12: 0.1012 worse than 0.117, patience: 2/3!


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.1334, 'crossentropy': 3.7211234298706053}


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.1526, 'crossentropy': 3.6456055068969726}


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.1506, 'crossentropy': 3.639182250213623}
Epoch 15: 0.1506 worse than 0.1526, patience: 1/3!


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.1556, 'crossentropy': 3.5743414745330813}


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.1798, 'crossentropy': 3.5142369106292723}


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.1766, 'crossentropy': 3.4968504013061525}
Epoch 18: 0.1766 worse than 0.1798, patience: 1/3!


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.19, 'crossentropy': 3.415262245178223}


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.2, 'crossentropy': 3.436680059814453}


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.2038, 'crossentropy': 3.479621842575073}


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.1874, 'crossentropy': 3.492301457977295}
Epoch 22: 0.1874 worse than 0.2038, patience: 1/3!


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.2194, 'crossentropy': 3.321240447998047}


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.2264, 'crossentropy': 3.2894559776306154}


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.2468, 'crossentropy': 3.2592536624908446}


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.255, 'crossentropy': 3.1534479541778566}


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.2574, 'crossentropy': 3.194382339859009}


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.2588, 'crossentropy': 3.2088769870758056}


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.2538, 'crossentropy': 3.1996145584106443}
Epoch 29: 0.2538 worse than 0.2588, patience: 1/3!


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.2868, 'crossentropy': 3.048496583557129}


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.285, 'crossentropy': 3.0462064266204836}
Epoch 31: 0.285 worse than 0.2868, patience: 1/3!


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.2916, 'crossentropy': 3.0441914291381837}


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.3042, 'crossentropy': 2.9938636936187746}


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.3142, 'crossentropy': 2.9616074798583982}


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.3128, 'crossentropy': 3.0372639751434325}
Epoch 35: 0.3128 worse than 0.3142, patience: 1/3!


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.3266, 'crossentropy': 2.9172872680664064}


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.332, 'crossentropy': 2.8956572322845457}


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.3346, 'crossentropy': 2.8784393520355223}


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.3478, 'crossentropy': 2.834599676513672}


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.3372, 'crossentropy': 2.8797668670654297}
Epoch 40: 0.3372 worse than 0.3478, patience: 1/3!


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.3616, 'crossentropy': 2.7556054058074952}


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.3524, 'crossentropy': 2.8509595169067383}
Epoch 42: 0.3524 worse than 0.3616, patience: 1/3!


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.3544, 'crossentropy': 2.8260359642028807}
Epoch 43: 0.3544 worse than 0.3616, patience: 2/3!


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.3708, 'crossentropy': 2.761400556564331}


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.3746, 'crossentropy': 2.743904900741577}


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.3752, 'crossentropy': 2.752844585418701}


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.391, 'crossentropy': 2.7870617553710937}


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.3912, 'crossentropy': 2.6811590377807617}


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.392, 'crossentropy': 2.7096566692352293}


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.3922, 'crossentropy': 2.6482392658233644}


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.4076, 'crossentropy': 2.69913618888855}


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.3926, 'crossentropy': 2.7104971210479736}
Epoch 52: 0.3926 worse than 0.4076, patience: 1/3!


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.4212, 'crossentropy': 2.7271564582824706}


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.3974, 'crossentropy': 2.7731422164916992}
Epoch 54: 0.3974 worse than 0.4212, patience: 1/3!


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.4226, 'crossentropy': 2.7127608409881594}


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.4286, 'crossentropy': 2.686399506378174}


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.4234, 'crossentropy': 2.731360105895996}
Epoch 57: 0.4234 worse than 0.4286, patience: 1/3!


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.4368, 'crossentropy': 2.6142911724090574}


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.4238, 'crossentropy': 2.699748579788208}
Epoch 59: 0.4238 worse than 0.4368, patience: 1/3!


[1/141]   1%|           [00:00<?]

[1/40]   2%|2          [00:00<?]

Epoch metrics: {'accuracy': 0.4228, 'crossentropy': 2.831049083137512}
Epoch 60: 0.4228 worse than 0.4368, patience: 2/3!


AdamW (
Parameter Group 0
    amsgrad: False
    betas: (0.9, 0.999)
    eps: 1e-08
    lr: 0.005
    weight_decay: 0.0005
)

In [ ]:
# experiment

evaluate(model=model_optimizer.model, num_samples=20, loader=test_loader, device="cuda")

[1/79]   1%|1          [00:00<?]

{'accuracy': 0.5724, 'crossentropy': 3.0206112657546997}

57.24% is better than Random acquisition in VAAL (Variational Adversarial Active Learning, https://arxiv.org/abs/1904.00370) and also better than the reported CoreSet AL results: 50.5%?

---

"Task-Aware Variational Adversarial Active Learning" (https://arxiv.org/abs/2002.04709) reports much better numbers for Random on ResNet, however: 64%.

They state:

> ResNet18 [13] was used for all task learners and stochastic gradient descent (SGD) was used with momentum 0.9 and weight decay 0.005. Learning rate was 0.1 for the first 160 epochs and then 0.01 for the last 40 epochs.

We leave this as a TODO.